File: Chapter-05-Linked-Lists/List-Contains/List-Contains.py
# Write a function to check if a value is in any list node or not

# node class
class Node(object):
    def __init__(self, value):
        self.val = value
        self.next = None

# linked list class
class lList(object):
    def __init__(self):
        self.__head = None
        self.__tail = None
        self.__length = 0

    def addFront(self, val):
        newNode = Node(val)

        if not self.__head:
            self.__head = newNode
            self.__tail = newNode
        else:
            temp = self.__head
            self.__head = newNode
            newNode.next = temp

        self.__length += 1

    def removeFront(self):
        if not self.__head:
            front = "Empty List"
        else:
            front = self.__head.val
            if self.__head == self.__tail:
                self.__head = None
                self.__tail = None
            else:
                self.__head = self.__head.next

            self.__length -= 1

        return front

    def remove(self):
        if not self.__head:
            back = "Empty List"
        else:
            current = self.__head
            back = self.__tail.val
            if current == self.__tail:
                self.__head = None
                self.__tail = None
            else:
                while current.next != self.__tail:
                    current = current.next
                current.next = None
                self.__tail = current

            self.__length -= 1

        return back

    def add(self, val):
        newNode = Node(val)

        if not self.__tail:
            self.__head = newNode
            self.__tail = newNode
        else:
            self.__tail.next = newNode
            self.__tail = newNode

        self.__length += 1

    def length(self):
        return self.__length

    def show(self):
        if not self.__head:
            print "Empty List"
        else:
            result = ''
            current = self.__head
            while current:
                result = "{}{} ".format(result, current.val)
                current = current.next

            print result

    def search(self, value):
        if self.__head:
            current = self.__head
            while current:
                if current.val == value:
                    return True
                current = current.next
            return False
        else:
            return False

# create the list
newList = lList()
newList.show()
print newList.length()

# populate the list
for i in range(1, 11):
    newList.add(i)
newList.show()
print newList.length()

myVal = 2
# search for myVal in the list
if newList.search(myVal):
    print "{} is in the list".format(myVal)
else:
    print "{} is not in the list".format(myVal)

File: Chapter-05-Linked-Lists/List-Contains/List-Contains.js
// Write a function to check if a value is in any list node or not

// node class
function Node(value){
  this.val = value
  this.next = null
}

// linked list class
function lList(){
  var head = null
  var tail = null
  var length = 0

  this.addFront = function(val){
    var newNode = new Node(val)
    if (!head){
      head = newNode
      tail = newNode
    } else {
      var temp = head
      head = newNode
      newNode.next = temp
    }
    length++
  }

  this.add = function(val){
    var newNode = new Node(val)
    if (!tail){
      head = newNode
    } else {
      tail.next = newNode
    }
    tail = newNode
    length++
  }

  this.removeFront = function(){
    if (!head){
      var front = "Empty List"
    } else {
      var front = head.val
      if (head == tail){
        head = null
        tail = null
      } else {
        head = head.next
      }
      length--
    }
    return front
  }

  this.remove = function(){
    if (!head){
      var back = "Empty List"
    } else {
      var back = tail.val
      var current = head

      if (current == tail){
        head = null
        tail = null
      } else {
        while (current.next != tail){
          current = current.next
        }
        tail = current
      }
      length--
    }
    return back
  }

  this.length = function(){
    return length
  }

  this.show = function(){
    if (!head){
      console.log("Empty List")
    } else {
      var result = ''
      var current = head
      while (current){
        result = result + current.val + ' '
        current = current.next
      }
      console.log(result)
    }
  }

  this.search = function(value){
    if (head){
      var current = head
      while (current){
        if (current.val == value){
          return true
        }
        current = current.next
      }
      return false
    } else {
      return false
    }
  }
}

// create a list first
var newList = new lList()
newList.show()
console.log(newList.length())

// populate the list
for (let i = 1; i < 11; i++){
  newList.add(i)
}
newList.show()
console.log(newList.length())

var myVal = 11
// search for myVal in the list
if (newList.search(myVal)){
  console.log(`${myVal} is in the list`)
} else {
  console.log(`${myVal} is not in the list`)
}
